lib/analyse.ml

```diff
@@ -173,6 +178,44 @@ module Analysis = struct
       ) OpamPackage.Map.empty
     >|= Result.ok
 
+  let has_tests opam =
+    let has_with_test_variable =
+      let exception With_test_found in
+      let aux variable =
+        if OpamVariable.Full.is_global variable then
+          match OpamVariable.to_string (OpamVariable.Full.variable variable) with
+          | "with-test" -> raise With_test_found
+          | _ -> variable
+        else
+          variable
+      in
+      try
+        let _ : OpamFile.OPAM.t =
+          OpamFileTools.map_all_variables aux opam
+        in
+        true
+      with With_test_found -> true
+    in
+    match OpamFile.OPAM.run_test opam with
+    | [] -> has_with_test_variable
+    | _::_ -> true
+
+  let package_to_path pkg =
+    let name = OpamPackage.Name.to_string (OpamPackage.name pkg) in
+    let version = OpamPackage.Version.to_string (OpamPackage.version pkg) in
+    let ( // ) = Fpath.( / ) in
+    Fpath.v "packages" // name // (name^"."^version) // "opam"
+
+  let map_has_tests ~dir (pkg, kind) =
+    let path = Fpath.to_string (package_to_path pkg) in
+    get_opam ~cwd:dir path >|= function
+    | Error () -> assert false
+    | Ok content ->
+        let filename = OpamFile.make (OpamFilename.raw path) in
+        let content = OpamFile.OPAM.read_from_string ~filename content in
+        let has_tests = has_tests content in
+        (pkg, {kind; has_tests})
+
```

Comment:
@rjbou could you have a look at this code to double-check there isn't something i forgotten? This code is suppose to check if a package has tests or not.
Is checking if the package contains `with-test` or has a non-empty `run-test` field enough? Is the `with-test` variable not substituted during parsing?

Reply:
You have full coverage, no `with-test` added, only resolved for filtering.
You don't use `OpamFileTools.all_variables` for optimisation purpose?

Reply:
Cool. `OpamFileTools.all_variables` is not exported so I can't use it here.

Reply:
oh ok